RFC-0004-pyTorch-sparse-matmul-roadmap.md

```diff
@@ -0,0 +1,544 @@
+
+Among projects tailored to neural networks & deep learning,
+[Tensorflow](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseMatrixSparseMatMul)
+has a `SparseTensor` class that uses COO format for tensors of arbitrary
+rank `n` and a `SparseMatrix` class that uses CSR format for tensors of
+rank 2; explicit conversion routines between formats exist where they apply. 
+There are also rank 2 `CSRNDArray` and rank 1 `RowSparseMatrix` classes 
+included in
+[MXNet](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseMatrixSparseMatMul).
+The project [`pytorch_sparse`](https://github.com/rusty1s/pytorch_sparse)
+suggests some directions for supporting efficient sparse matrix operations
+with `autograd` and also with CUDA support.
+
+Finally, there are generalizations of sparse matrix storage formats to sparse 
+tensors of arbitrary rank `n` described in various articles listed in the 
+references.
+
+## Implementation
+
+The module `torch` already contains an attribute `torch.sparse_coo` of type
+`torch.layout` with an associated constructor `torch.sparse_coo_tensor`. The
+proposed CSR implementation should extend the API by adding a few attributes
+to the `torch` module with similar names:
+
++ `torch.sparse_csr` (an attribute of type `torch.layout`);
++ `torch.sparse_csr_tensor` (an associated constructor); and
++ functions `torch.to_sparse_csr` & `torch.to_sparse_coo` for explicit
+  conversion between representations (for rank 2 sparse tensors).
+
+The `torch.sparse_csr_tensor` constructor would have a similar signature to `torch.sparse_coo_tensor`, i.e.,
+
+```{python}
+sparse_csr_tensor(indices, values, size=None, dtype=None,
+                  device=None, requires_grad=False)       -> Tensor
+```
+
+where `indices` would be a list of two lists (e.g., `RO` & `CO`
+from the description of CSR format above) and `values` would be a tensor
+containing the corresponding (presumably) nonzero entries of the sparse
+tensor. This is sufficiently general to account for tensors of rank `d>2`,
+thereby enabling batches of sparse matrices. It should also be acceptable
+for `values` to contain tensors (of fixed dimensions) rather than scalars.
+
+The basic outline, then, is as follows:
+
+1. Construct an exhaustive suite of benchmark tests to time the existing
+   implementation of sparse matrix-vector & matrix-matrix products in PyTorch.
+
+   + See [issue # 16187](https://github.com/pytorch/pytorch/issues/16187)
+     for representative examples of benchmarks.
+   + Include examples of matrices of fairly large dimensions (i.e., where dense
+     representations are impractical or infeasible).
+   + Include examples with `sparse x dense` and `sparse x sparse` matrices for
+     the timings.
+
+2. Construct and test a prototype implementation of conversion functions
+   `torch.to_sparse_csr` and `torch.to_sparse_coo` to convert sparse PyTorch
+   tensors of sparse rank 2 between COO and CSR formats.
```

Comment:
I think we can target immediately arbitrary rank (>=2) CSR sparse tensors in the constructor, including the hybrid feature, without any extra effort. The algorithms, however, may be restricted to 2-D in the short run.